torch/csrc/jit/python/python_sugared_value.cpp

```diff
@@ -252,6 +254,27 @@ SugaredValuePtr ModuleValue::getitem(
         }
       }
       throw ErrorReport(loc) << "Key Error, " << idx_str;
+    } else if (type_hint) {
+      // Check that all submodules comply with the type hint.
+      const auto& self_type = concreteType_->getJitType()->expect<ClassType>();
+      for (size_t i = 0; i < self_type->numAttributes(); ++i) {
+        const auto& attr_type = self_type->getAttribute(i);
+        if (attr_type->is_module()) {
+          if (!attr_type->isSubtypeOf(type_hint)) {
+            auto loc = self_->node()->sourceRange();
+            throw ErrorReport(loc)
+                << "Attribute " << self_type->getAttributeName(i)
+                << " is not of annotated type " << type_hint->annotation_str();
+          }
+        }
+      }
+
+      // Emit a prim::ModuleDictIndex operator.
+      auto graph = m.graph();
```

Comment:
We should add clarification to why we’re making it a built in op. 
I think it’s bc: there’s no good way of constructing a dict representing the ModuleDict and caching it across indexes, bc any call to setAttr on the module would invalidate the dict, and we want to preserve the O(1) complexity